src/Raven.Server/Web/RequestHandler.cs

```diff
@@ -573,18 +574,57 @@ private static void ThrowInvalidAuthStatus(RavenServer.AuthenticationStatus? sta
             throw new ArgumentOutOfRangeException("Unknown authentication status: " + status);
         }
 
-        public static void SetupCORSHeaders(HttpContext httpContext)
+        public static void SetupCORSHeaders(HttpContext httpContext, ServerStore serverStore)
         {
+            httpContext.Response.Headers.Add("Vary", "Origin");
+            
+            var requestedOrigin = httpContext.Request.Headers["Origin"];
 
-            httpContext.Response.Headers.Add("Access-Control-Allow-Origin", httpContext.Request.Headers["Origin"]);
+            string allowedOrigin = null; // prevent access by default
+            
+            if (IsOriginAllowed(requestedOrigin, serverStore))
+            {
+                allowedOrigin = requestedOrigin;
+            }
+            
+            httpContext.Response.Headers.Add("Access-Control-Allow-Origin", allowedOrigin);
             httpContext.Response.Headers.Add("Access-Control-Allow-Methods", "PUT, POST, GET, OPTIONS, DELETE");
             httpContext.Response.Headers.Add("Access-Control-Allow-Headers", httpContext.Request.Headers["Access-Control-Request-Headers"]);
             httpContext.Response.Headers.Add("Access-Control-Max-Age", "86400");
         }
 
+        private static bool IsOriginAllowed(string origin, ServerStore serverStore)
+        {
+            if (serverStore.Server.Certificate.Certificate == null)
+            {
+                // running in unsafe mode - since server can be access via multiple urls/aliases accept them 
+                return true;
+            }
+            
+            var topology = serverStore.GetClusterTopology();
```

Comment:
Do we need to worry about case sensitivity here?

Reply:
Good question, I will check that using empirical mode. :)

Reply:
It needs to be case insensitive. Thanks. I will fix that.